src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form_is_modified.ts

```diff
@@ -81,23 +83,27 @@ export const useFormIsModified = ({
     .filter(isFieldIncluded)
     .some(([_, field]) => field.isModified);
 
-  if (isModified) {
-    return isModified;
-  }
+  if (!isModified) {
+    // 2. Check if any field has been removed.
+    // If somme field has been removed **and** they were originaly present on the
+    // form "defaultValue" then the form has been modified.
+    const formDefaultValue = __getFormDefaultValue();
+    const fieldOnFormDefaultValue = (path: string) => Boolean(get(formDefaultValue, path));
 
-  // 2. Check if any field has been removed.
-  // If somme field has been removed **and** they were originaly present on the
-  // form "defaultValue" then the form has been modified.
-  const formDefaultValue = __getFormDefaultValue();
-  const fieldOnFormDefaultValue = (path: string) => Boolean(get(formDefaultValue, path));
+    const fieldsRemovedFromDOM: string[] = fieldsToDiscard
+      ? Object.keys(__getFieldsRemoved())
+          .filter((path) => fieldsToDiscard[path] !== true)
+          .filter(fieldOnFormDefaultValue)
+      : Object.keys(__getFieldsRemoved()).filter(fieldOnFormDefaultValue);
 
-  const fieldsRemovedFromDOM: string[] = fieldsToDiscard
-    ? Object.keys(__getFieldsRemoved())
-        .filter((path) => fieldsToDiscard[path] !== true)
-        .filter(fieldOnFormDefaultValue)
-    : Object.keys(__getFieldsRemoved()).filter(fieldOnFormDefaultValue);
+    isModified = fieldsRemovedFromDOM.length > 0;
+  }
 
-  isModified = fieldsRemovedFromDOM.length > 0;
+  if (isModified && !isFormModified) {
```

Comment:
The `isModified` is the new derived state calculated on every render and `isFormModified` is the actual state that is returned to the consumer.

Ideally instead of 

```js
if (isModified && !isFormModified) {
  setIsFormModified(true);
} else if (!isModified && isFormModified) {
  setIsFormModified(false);
}
```

I wanted to have just 

```js
setIsFormModified(isModified);
```

But that was creating an infinite state update and re-render.

I've renamed the `isModified` to `nextIsModified` to make it clear + added a comment